pallets/gear/src/lib.rs

```diff
@@ -826,10 +832,12 @@ pub mod pallet {
                 .unwrap_or_default()
         }
 
-        /// Returns true if there is a program with the specified id (it may be paused).
+        /// Returns true if there is a program with the specified `program_id`` (it may be paused)
+        /// or this `program_id` belongs to the built-in actor.
         pub fn program_exists(program_id: ProgramId) -> bool {
             ProgramStorageOf::<T>::program_exists(program_id)
                 || ProgramStorageOf::<T>::paused_program_exists(&program_id)
+                || T::BuiltinRouter::provide().lookup(&program_id).is_some()
```

Comment:
btw it's conceptually not a program, but actor, so consider removing here as well as above

Reply:
The `program_exists` function must be aware of the Builtin destinations. Agree, that "providing" every time is sub-optimal, but I don't see how we could remove this check. Because all "handle" messages destinations have to meet this requirement, do they?